main/main_timer_sync.cpp

```diff
@@ -264,5 +264,20 @@ void MainTimerSync::set_fixed_fps(int p_fixed_fps) {
 MainFrameTime MainTimerSync::advance(double p_physics_step, int p_physics_ticks_per_second) {
 	double cpu_process_step = get_cpu_process_step();
 
-	return advance_checked(p_physics_step, p_physics_ticks_per_second, cpu_process_step);
+	MainFrameTime mft = advance_checked(p_physics_step, p_physics_ticks_per_second, cpu_process_step);
+
+	// Now backcalculate the logical timing of the first physics tick.
+	// This is used for processing input.
+	// It is approximate, but should be fine for input.
+	mft.usec_per_tick = 1000000 / p_physics_ticks_per_second;
+	uint64_t leftover_usec = mft.interpolation_fraction * mft.usec_per_tick;
+
+	// Note we are using the ACTUAL CPU time for this estimate,
+	// NOT the smoothed accumulated time.
+	// This is because the input timestamps are measured in realtime,
+	// and smoothed time / realtime can get out of sync.
+	mft.first_physics_tick_logical_time_usecs = current_cpu_ticks_usec;
+	mft.first_physics_tick_logical_time_usecs -= (mft.physics_steps * mft.usec_per_tick) + leftover_usec;
+
```

Comment:
Just to note that the timing calculations for the first logical physics tick are more complex when delta smoothing is being used (as there is time compression and expansion compared to wall clock time). I'm just fixing this up for the 3.x PR, but delta smoothing isn't merged yet in 4.x so will leave the simple version here for now.